File: platforms/winui/src/MapData.cpp
#include "MapData.h"
#include "MapData.g.cpp"

#include "MapController.h"

namespace winrt::TangramWinUI::implementation {
uint64_t MapData::AppendOrUpdateGeoJson(array_view<uint8_t const>& value, bool generateTiles) {
    if (value.size() == 0) return 0;

    if (IsInvalid()) return 0;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return 0;
    
    auto result = m_source->appendOrUpdateFeatures(reinterpret_cast<const char*>(value.begin()), value.size());
    if (generateTiles) GenerateTilesLockFree();
    return result;
}

uint64_t MapData::RemoveGeoJsonById(array_view<uint64_t const>& ids, bool generateTiles) {
    if (ids.size() == 0) return 0;

    if (IsInvalid()) return 0;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return 0;

    auto result = m_source->removeFeatures(ids.begin(), ids.size());
    if (generateTiles) GenerateTilesLockFree();
    return result;
}

void MapData::IsVisible(bool isVisible) {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return;

    m_source->setVisible(isVisible);
}

bool MapData::IsVisible() const {
    if (IsInvalid()) return false;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return false;

    return m_source->isVisible();
}

bool MapData::CanUpdateFeatures() const {
    if (IsInvalid()) return false;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return false;

    return m_source->canUpdateFeatures();
}

void MapData::CanUpdateFeatures(bool enabled) {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return;

    m_source->setCanUpdateFeatures(enabled);
}

void MapData::Clear() {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return;

    m_source->clearFeatures();
    GenerateTilesLockFree();
}

void MapData::GenerateTiles() {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return;

    GenerateTilesLockFree();
}

void MapData::Init(Tangram::ClientDataSource* source, MapController* controller) {
    m_source = source;
    m_controller = controller;
    m_name = to_hstring(source->name());
    m_id = source->id();
}

void MapData::GenerateTilesLockFree() {
    auto &map = m_controller->GetMap();
    map.clearTileCache(m_id);
    m_source->generateTiles();
    m_controller->RequestRender();
}

bool MapData::IsInvalid() const {
    return !m_controller || m_controller->IsShuttingDown();
}

void MapData::SetGeoJson(const hstring& geoJson) {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return;

    m_source->clearFeatures();
    m_source->addData(reinterpret_cast<const char*>(geoJson.begin()), geoJson.size());

    GenerateTilesLockFree();
}

void MapData::SetGeoJsonFromBytes(array_view<uint8_t const>& value) {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if(IsInvalid()) return;
    
    m_source->clearFeatures();
    m_source->addData(reinterpret_cast<const char*>(value.begin()), value.size());

    GenerateTilesLockFree();
}

void MapData::Remove() {
    if (IsInvalid()) return;
    std::scoped_lock mapLock(m_controller->Mutex());
    if (IsInvalid()) return;
    
    m_controller->RemoveDataLayer(*this);
    Invalidate();
}

Tangram::ClientDataSource* MapData::Source() { return m_source; }

} // namespace winrt::TangramWinUI::implementation

File: platforms/android/tangram/src/main/java/com/mapzen/tangram/MapData.java
package com.mapzen.tangram;

import com.mapzen.tangram.geometry.Geometry;

import java.util.List;

import androidx.annotation.NonNull;

/**
 * {@code MapData} is a named collection of drawable map features.
 */
public class MapData {

    private MapController mapController;

    final String name;
    long pointer;

    /**
     * For package-internal use only; create a new {@code MapData}
     *
     * @param name    The name of the associated data source
     * @param pointer The markerId to the native data source, encoded as a long
     * @param map     The {@code MapController} associated with this data source
     */
    MapData(final String name, final long pointer, @NonNull final MapController map) {
        this.name = name;
        this.pointer = pointer;
        this.mapController = map;
    }

    /**
     * Assign a list of features to this data collection. This replaces any previously assigned feature lists or GeoJSON data.
     *
     * @param features The features to assign
     */
    public void setFeatures(@NonNull final List<Geometry> features) {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        nativeMap.clearClientDataFeatures(pointer);

        for (Geometry feature : features) {
            nativeMap.addClientDataFeature(pointer,
                    feature.getCoordinateArray(),
                    feature.getRingArray(),
                    feature.getPropertyArray());
        }
        nativeMap.generateClientDataTiles(pointer);
        mapController.requestRender();
    }

    public void setCanUpdateFeatures(boolean enabled)
    {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        nativeMap.setCanUpdateFeatures(pointer, enabled);
    }

    public boolean canUpdateFeatures()
    {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        return nativeMap.canUpdateFeatures(pointer);
    }


    /**
     * Assign features described in a GeoJSON string to this collection. This will replace any previously assigned feature lists or GeoJSON data.
     *
     * @param data A string containing a <a href="http://geojson.org/">GeoJSON</a> FeatureCollection
     */
    public void setGeoJson(final String data) {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;

        nativeMap.clearClientDataFeatures(pointer);
        nativeMap.addClientDataGeoJson(pointer, data);
        nativeMap.generateClientDataTiles(pointer);
        mapController.requestRender();
    }

    /**
     * Set the visibility of all the features in this {@code MapData}.
     *
     * @param visible If true, the features are drawn on the map. Otherwise they are hidden.
     */
    public void setVisible(boolean visible) {
        checkPointer(pointer);
        mapController.nativeMap.setClientDataVisible(pointer, visible);
    }

    /**
     * Get the current visibility of all the features in this {@code MapData}.
     *
     * @return If true, the features are currently drawn on the map. Otherwise, they are currently hidden.
     */
    public boolean getVisible() {
        checkPointer(pointer);
        return mapController.nativeMap.getClientDataVisible(pointer);
    }

    /**
     * Get the name of this {@code MapData}.
     *
     * @return The name.
     */
    public String name() {
        return name;
    }

    /**
     * Remove this {@code MapData} from the map it is currently associated with. Using this object
     * after {@code remove} is called will cause an exception to be thrown. {@code remove} is called
     * on every {@code MapData} associated with a map when its {@code MapController} is destroyed.
     */
    public void remove() {
        final MapController map = mapController;
        if (map == null) {
            return;
        }
        map.removeDataLayer(this);
    }

    void dispose() {
        mapController = null;
        pointer = 0;
    }

    /**
     * Remove all features from this collection.
     */
    public void clear() {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        nativeMap.clearClientDataFeatures(pointer);
        nativeMap.generateClientDataTiles(pointer);
        mapController.requestRender();
    }

    private void checkPointer(final long ptr) {
        if (ptr == 0) {
            throw new RuntimeException("Tried to perform an operation on an invalid pointer!"
                    + " This means you may have used a MapData that has already been removed.");
        }
    }

    /**
     * Assign features described in a GeoJSON UTF-8 byte array to this collection. This will replace any previously assigned feature lists or GeoJSON data.
     *
     * @param data A UTF-8 byte array containing a <a href="http://geojson.org/">GeoJSON</a> FeatureCollection
     */
    public void setGeoJsonFromBytes(final byte[] data) {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;

        nativeMap.clearClientDataFeatures(pointer);
        nativeMap.addClientDataGeoJsonFromBytes(pointer, data);
        nativeMap.generateClientDataTiles(pointer);
        mapController.requestRender();
    }

    public void generateTiles()
    {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        nativeMap.generateClientDataTiles(pointer);
        mapController.requestRender();
    }

    private void generateTilesIfNeeded(NativeMap nativeMap, long count, boolean generateTiles)
    {
        // count != 0 cares about if we somehow managed to overflow from unsigned uint64_t
        // but it is not really possible.
        if(count != 0 && generateTiles) {
            nativeMap.generateClientDataTiles(pointer);
            mapController.requestRender();
        }
    }

    public long appendOrUpdateGeoJson(final byte[] data, boolean generateTiles)
    {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        long count = nativeMap.appendOrUpdateClientDataFromBytes(pointer, data);
        generateTilesIfNeeded(nativeMap, count, generateTiles);
        return count;
    }

    public long removeGeoJsonById(final long[] ids, boolean generateTiles)
    {
        checkPointer(pointer);
        final NativeMap nativeMap = mapController.nativeMap;
        long count = nativeMap.removeClientDataById(pointer, ids);
        generateTilesIfNeeded(nativeMap, count, generateTiles);
        return count;
    }
}